### PR TITLE
channel binding

### PR DESCRIPTION
**Pull request by conradludgate**
## Problem

channel binding protects scram from sophisticated MITM attacks where the attacker is able to produce 'valid' TLS certificates.

## Summary of changes

get the tls-server-end-point channel binding, and verify it is correct for the SCRAM-SHA-256-PLUS authentication flow

## Checklist before requesting a review

- [X] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: proxy/src/stream.rs [hunk 1]
```diff
@@ -118,19 +119,21 @@ impl<S: AsyncWrite + Unpin> PqStream<S> {
...
+pub enum Stream<S> {
+    /// We always begin with a raw stream,
+    /// which may then be upgraded into a secure stream.
+    Raw { raw: S },
+    /// We box [`TlsStream`] since it can be quite large.
+    Tls {
+        tls: Box<TlsStream<S>>,
+        certified_key: Option<Arc<CertifiedKey>>,
+    },
 }
```
Comment (hlinnaka) [inline proxy/src/stream.rs:133]: This pinning change seems like an independent change. Maybe extract that to a separate commit?

Comment (hlinnaka) [reply proxy/src/stream.rs:133]: BTW, what does "decomposing this object as necessary" mean?

File: proxy/src/stream.rs [hunk 2]
```diff
@@ -4,6 +4,7 @@ use bytes::BytesMut;
 use pin_project_lite::pin_project;
```
Comment (hlinnaka) [inline proxy/src/stream.rs:4]: this 'use' is now unused

**Review by hlinnaka — COMMENTED** [review]
Nice! I thought there was some fundamental reason we couldn't do channel binding in the proxy - happy to see that we can!

**Comment by conradludgate** [comment]
There still might be 😅

File: proxy/src/stream.rs (hunk 2 above)
Comment (conradludgate) [reply proxy/src/stream.rs:4]: I even get nitpicks in my draft PRs 😭

File: proxy/src/stream.rs (hunk 1 above)
Comment (conradludgate) [reply proxy/src/stream.rs:133]: Sure, I can split the commit.

> BTW, what does "decomposing this object as necessary" mean?

Not sure, I'd have to look up the history of that comment I think

**Comment by conradludgate** [comment]
Local testing seems to work. I will try and deploy it to preview and test further

**Comment by conradludgate** [comment]
Working in preview :D

**Review by khanova — APPROVED** [review]
Looks good from my perspective.

File: proxy/src/config.rs [hunk 3]
```diff
@@ -76,35 +79,89 @@ pub fn configure_tls(
...
+        let sha256_oids = [
+            oid_registry::OID_SIG_ECDSA_WITH_SHA256,
+            oid_registry::OID_PKCS1_SHA256WITHRSA,
+        ];
+
+        let pem = x509_parser::parse_x509_certificate(&cert.0)
+            .context("Failed to parse PEM object from cerficiate")?
+            .1;
+
+        if sha256_oids.contains(pem.signature_algorithm.oid()) {
```
[the remark this answers is not in the document]
Comment (conradludgate) [reply proxy/src/config.rs:139]: Yeah this is a complicated one. We should check for md5 or sha1, but also I really hope we never use certificates that use these signature algorithms 😅. In my experiments, I noticed only the two signature algorithms above.

I think what I will do is on startup log the oids of the certificates and raise a warning if they don't match expectations.

File: proxy/src/auth/flow.rs [hunk 4]
```diff
@@ -119,13 +135,19 @@ impl<S: AsyncRead + AsyncWrite + Unpin> AuthFlow<'_, S, Scram<'_>> {
             .ok_or(AuthErrorImpl::MalformedPassword("bad sasl message"))?;
 
         // Currently, the only supported SASL method is SCRAM.
-        if !scram::METHODS.contains(&sasl.method) {
+        if !scram::METHODS_PLUS.contains(&sasl.method) {
```
[the remark this answers is not in the document]
Comment (conradludgate) [reply proxy/src/auth/flow.rs:138]: The `METHOD` case is left as a fallback in case we can't determine the certificate hash for channel binding.

File: proxy/src/config.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (khanova) [reply proxy/src/config.rs:139]: That makes sense!

**Comment by Neustradamus** [comment]
@conradludgate: Good job!

Linked to:
- https://github.com/scram-sasl/info/issues/1
